### DISCUSSION TITLE
Resize image according to blur

### DISCUSSION THREAD

**Discussion by Atomiz2002** [ImageMagick/ImageMagick#6752] [Help]
### ImageMagick version

7.1.1-19

### Operating system, version and so on

Win 10

### Description

How can I blur an image without cropping out the leftover blurred pixels, but instead resizing the image to contain them with aspect ratio preserved?

(I've thought of enlarging the image by a few times the blur amount, blurring it, trimming it, and extending it to revert the ratio, but this method seems both inefficient and very complex to execute)

**Comment by snibgo** [comment]
I don't understand the question.

What are "the leftover blurred pixels"?

How would "resizing the image" do what you want?

**Reply by Atomiz2002** [reply]
Say I have an image with dimensions 500x500, and a square inside of it with 450x450. If I apply gaussian blur on it the full blurred image would not fit in the initial dimensions and the "leftover pixels" would not be displayed, it would require quite a bit more space to contain the full blurred image so that it doesn't have an cutoff such as 
![image](https://github.com/ImageMagick/ImageMagick/assets/24458572/7841ab3d-9abc-4dac-81e0-b95f17969f86)

**Comment by snibgo** [comment]
> Say I have an image with dimensions 500x500, and a square inside of it with 450x450.

What does that mean? Perhaps the 500x500 image is one colour such as black, except for a 450x450 square within the image that is a different colour such as green.

Sorry, I still don't understand. If your image is 500x500 pixels and you blur it, the result will have 500x500 pixels. Pixels that are outside the 500x500, "virtual pixels", are not blurred. If the image had been larger than 500x500, perhaps some would have become green.

So perhaps your question really is, "If I apply a "-blur 0x50" etc, how far will the effect spread?" That question is easily answered by starting with a much larger image, and trimming it.
```
magick -size 450x450 xc:Lime -bordercolor Black -border 1000 -blur 0x50 -trim +repage info:

xc:Lime XC 800x800 16-bit sRGB 0.000u 0:00.000
```
The blur has spread greenness to the boundary of a square that is 800x800 pixels, or 175 extra pixels to the top, left, right and bottom of the 450x450 square. This result will vary with your Q-number, and the contrast between the two colours.

**Reply by Atomiz2002** [reply]
the background of the image is transparent, the preview is black. starting with a border isn't optimal as it's a relatively heavy operation when done a dozen times which is what I'm after so my question is if there is a way to automatically let the image be resized to fit in all the blur done to it without cutting off pixels and the aspect ratio to be preserved or perhaps it might be a feature request at this point if non existent already. the border is a half working workaround as it wouldn't preserve the aspect ratio of the image. I've tried using resize but I'm struggling to find an optimal set of values to fit both small and large images with both small and large blurs

**Reply by fmw42** [reply]
I note that a Gaussian blur extends forever. I never gets to zero.  So you will alway cut it off with a finite extent of black around the image.

**Comment by snibgo** [comment]
A Gaussian blur theoretically extends to infinity, but IM will limit the kernel radius to where the value is more than 1/QuantumLevel. For example:
```
magick xc: -define morphology:showKernel=1 -morphology Convolve:0 Gaussian:0x10 null:

Kernel "Gaussian" of size 61x61+30+30 with values from 0 to 0.00159883
```

> I've tried using resize but ...

Please show what commands you have tried. I don't understand how `-resize` helps you.

**Reply by Atomiz2002** [reply]
I feel like you keep overlooking the fact i need the aspect ratio retained. I cant use border as it would ruin the ratio and i cant use trim as it would ruin the ratio, id need to identify the ratio, then extend the border which might be unnecessary if the sprite itself is smaller than the canvas and the blur doesnt even go beyond its borders, then trim it and then calculate how much i should extend it in the right direction to revert the image to its original ratio. Thats why i use resize to uniformly resize the image and keep the ratio.

As to what commands I have tried, not too many. Furthermore I'm using the blur as a filter on a downscaled image as my use case allows the images to be sometimes smaller and lower quality and sometimes bigger and higher quality in which case I don't apply any downscaling. If I recall correctly (because I'm not in front of my work now) I'm currently using -extent and not resize to account for the blur which would otherwise be cut off by the dimensions of the image.

**Comment by snibgo** [comment]
The command I showed doesn't change the aspect ratio. The input and output images are square.

> id need to identify the ratio, then extend the border which might be unnecessary if the sprite itself is smaller than the canvas and the blur doesnt even go beyond its borders, then trim it and then calculate how much i should extend it in the right direction to revert the image to its original ratio.

Yes, that is correct.

If the input isn't square, you can find what the trimmed dimensions would be, and adjust the numbers so the output is the same aspect ratio, and crop to that.

For example, suppose your input green rectangle is 300x200, "landscape" format. The "`-blur 0x50`" would add the same amount to each side, 175 pixels. We need to add extra pixels to the left and right to maintain the aspect ratio.

Windows BAT script.
```
set inW=300
set inH=200

for /F "usebackq tokens=1-2 delims=x+" %%A in (`%IMG7%magick ^
  -size %inW%x%inH% xc:Lime ^
  -bordercolor Black -border "%%[fx:max(w,h)]" +repage ^
  -blur 0x50 ^
  +write tmp.miff ^
  -format %%@ info:`) do (
  set WW=%%A
  set HH=%%B
)

echo Trim would be %WW%x%HH%

set /A newW=%HH%*%inW%/%inH%

echo Crop is %newW%x%HH%

%IMG7%magick ^
  tmp.miff ^
  -gravity Center ^
  -crop %newW%x%HH%+0+0 +repage ^
  out.png
```
If you also have "portrait" format inputs, you need to first check for this, and alter the height instead of the width.

If that doesn't do what you want, please provide specific example inputs.

**Reply by Atomiz2002** [reply]
this either seems or is way overcomplicated, im not in any way advanced and half the symbols look random to me. there is no way id be able to even test this. ~~i got the file with the content you provided but obviously it directly closes on run and i cant see how to insert any kind of image to the args so what am i supposed to do with it exactly?~~

i ran it on a stripped down version of imagemagick, thats my bad, let me check the results and investigate if ill be able to understand at least something at all,

**Comment by snibgo** [comment]
The only "random" part of my script is the input image. You haven't provided one, so I invented one.

[EDIT: You also didn't say what blur you wanted, so I used "-blur 0x50".]

Perhaps I misunderstand what you want. Please supply sample inputs, and the outputs you desire.
